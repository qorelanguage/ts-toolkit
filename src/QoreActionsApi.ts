export interface IQoreAppShared {
  display_name: string;
  short_desc: string;
  desc?: string;
}
export interface IQoreApp extends IQoreAppShared {
  name: string;
}

export interface IQoreAppAction extends IQoreAppShared {
  app: string;
  action: string;
  action_code: 1 | 2; // What are other possible values?
}

export type TQoreAppActionFunction = (obj?: string | Record<string, any>) => any;

export interface IQoreAppActionWithFunction extends IQoreAppAction {
  action_code: 2
  app_function: TQoreAppActionFunction;
}

export class QoreAppActionsApi {

}