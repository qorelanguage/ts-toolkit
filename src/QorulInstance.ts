/**
 * An config class to setup the config for the api 
 *
 * @category Model
 */
class QorusInstance {
    #id;

    /**Qore Technologies endpoint to authenticate the user */
    protected _endpoint; 

    /**Version of the api */
    protected _version;

    constructor(endpoint: string, version?: string){
        this._endpoint = endpoint;
        this._version = version ? version: 'latest';
    }

    /**
     * A getter to get the current authentication endpoint
     */
    get endpoint(){
        return this._endpoint;
    }

    /**
     * A setter to configure the Authentication endpoint
     */
    set endpoint(endpnt: string){
        this._endpoint = endpnt;
    }

    /**
     * A setter to configure the api version
     * 
     * @param version version of the api v...
     */
    set version(version: string){
        this._version = version;
    }

    /**
     * A getter to get the current version of the api to be used
     */
    get version(){
        return this._version;
    }

    /**
     * A getter to get the id of the instance
     */
    get id(){
        return this.#id;
    }
}

export default QorusInstance;