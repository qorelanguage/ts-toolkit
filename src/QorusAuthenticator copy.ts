import { getKeyValLocal, setKeyValLocal } from './managers/LocalStorage';
import logger from './managers/logger';
import { ApiPaths, apiPathsInitial, createApiPaths, Version, WithEndpointVersion } from './utils/apiPaths';
import { QorusRequest } from './QorusRequest';
import { AxiosError, AxiosResponse } from 'axios';

export interface Authenticator {
  /**
   * -login-function Takes optional username and password parameters to authenticate the user.
   If the username and password is not provided it tries to authenticate using the locally stored token from the selected {@link Endpoint}

   * token if the authentication is successful else returns undefined
   */
  login: (loginConfig: LoginParams) => Promise<string | undefined>;

  /**
   * -logout-function Logs out the current user from the selected endpoint
   *
   */
  logout: () => Promise<boolean>;

  /**
   * -initEndpoint-function Allows the user to add/initialize a new endpoint
   * @param props {@link InitEndpoint} requires url and accepts optional user and pass parameters to initialize the endpoint and then authenticate the user
   *
   * Returns the newly created endpoint
   */
  initEndpoint: (props: InitEndpoint) => Promise<Endpoint>;

  /**
   * -selectEndpoint-function Allows the user to select a endpoint from the endpoints array, logout the user from the current
   * selected endpoint
   *
   * @param id Id of the endpoint
   *
   * Returns endpoint if the operation is successful false otherwise.
   */
  selectEndpoint: (id: string) => Promise<Endpoint | undefined>;

  /**
   * -getSelectedEndpoint-function A getter to return selected {@link Endpoint}
   *
   * Returns the selected endpoint if it's created or returns undefined
   */
  getSelectedEndpoint: () => Endpoint | undefined;

  /**
   * -renewSelectedEndpointToken-function Allows the user to renew the selected endpoint authentication token
   *
   * @param props {@link LoginParams} optional username and password can be provided
   * Returns token if the authentication is successful, undefined otherwise
   */
  renewSelectedEndpointToken: (props: LoginParams) => Promise<string | undefined>;

  /**
   * -getEndpointById-function A getter to return the endpoint if it exist in the endpoints array
   *
   * @param id ID of the endpoint ex: "rippy"
   *
   * Returns the endpoint {@link Endpoint} if the endpoint with the supplied id exist in the endpoints array, undefined otherwise.
   */
  getEndpointById: (id: string) => Endpoint | undefined;

  /**
   * -setEndpointUrl-function A setter to set the url of the selected {@link Endpoint}
   * @param url Base url for the endpoint
   * @param id Optional id parameter to change the url of a particular endpoint
   *
   * Returns url of the endpoint if the operation is successful, undefined otherwise
   */
  setEndpointUrl: (url: string, id?: string) => Promise<string | undefined>;

  /** -setEndpointVersion-function A setter to edit the {@link Version} of the {@link Endpoint}
   * @param version Version of the qorus api
   * @param id Optional id parameter to change the url of a particular endpoint from the endpoints array
   *
   * Returns version of the endpoint if the operation is successful, undefined otherwise
   *
   */
  setEndpointVersion: (version: Version, id?: string) => Promise<Version | undefined>;

  /**
   * -getAuthToken-function A getter to return the auth token of the selected {@link Endpoint}
   *
   * Returns token if the the selected endpoint exists and the user is authenticated, otherwise returns undefined
   */
  getAuthToken: () => string | undefined;

  /**
   * -getEndpointVersion-function A getter to get the api {@link Version} of a {@link Endpoint}
   *
   * @param id Optional id parameter to get the version of a particular endpoint
   *
   * Returns version of the selected endpoint or version of the the endpoint found by id,
   * if the endpoint doesn't exists it returns undefined
   */
  getEndpointVersion: (id?: string) => Version | undefined;

  /**
   * -getApiPaths-function A getter to return the api paths for the selected {@link Endpoint}
   *
   * Returns ApiPaths for the selected endpoint if exists, otherwise returns default api paths
   */
  getApiPaths: () => ApiPaths;

  /**
   * -getAllEndpoints-function A getter to get all the available {@link Endpoint}
   *
   * Returns endpoints array with all the available endpoints
   */
  getAllEndpoints: () => Endpoint[];
}

export type QorusEndpointId = string;
export type QorusAuthToken = string;
export type QorusEndpointURL = string;

export interface WithQorusEndpointId {
  /**
   * Id for the {@link Endpoint} provided by the user, unique for every endpoint
   */
  id: QorusEndpointId;
}

export interface WithQorusAuthToken {
  /**
   * Auth token for the user provided {@link Endpoint}
   */
  authToken?: QorusAuthToken;
}

export interface WithQorusURL {
  /**
   * URL to your server for the provided {@link Endpoint}
   */
  url: QorusEndpointURL;
}

export interface LoginParams {
  /**
   * Username for the authentication to your server
   */
  user?: string;
  /**
   * Password for the authentication to your server
   */
  pass?: string;
}

export interface InitEndpoint extends WithQorusURL, WithEndpointVersion, WithQorusEndpointId, LoginParams {}

export interface Endpoint extends WithQorusURL, WithEndpointVersion, WithQorusAuthToken, WithQorusEndpointId {}

/**
 * Enables the user to authenticate with multiple user defined endpoints
 * @returns QorusAuthenticator object with all the supporting operations
 * @Category QorusAuthenticator
 */
const _QorusAuthenticator = (): Authenticator => {
  //** Array of user defined endpoints */
  const endpoints: Endpoint[] = [];

  /** All Api paths for the selected endpoint */
  let allApiPaths: ApiPaths = apiPathsInitial;

  let apiPathsAuthenticator: ApiPaths['authenticator'] = apiPathsInitial.authenticator;

  /** Selected endpoint from the endpoints array */
  let selectedEndpoint: Endpoint;

  let noauth = false;

  /**
   * A getter to return the endpoint if it exist in the endpoints array
   */
  const getEndpointById = (id: string): Endpoint | undefined => {
    return endpoints.find((endpoint) => endpoint.id === id);
  };

  /**
   * Logs out the current user from the selected endpoint
   */
  const logout = async (): Promise<boolean> => {
    if (selectedEndpoint) {
      let successful = false;

      try {
        await QorusRequest.post({ path: `${apiPathsAuthenticator.logout}` });

        selectedEndpoint.authToken = undefined;
        allApiPaths = apiPathsInitial;
        apiPathsAuthenticator = apiPathsInitial.authenticator;
        noauth = false;
        successful = true;
      } catch (error: any) {
        logger.log({
          level: 'error',
          message: `Couldn't logout user, ErrorCode: ${error.code}, ErrorMessage: ${error.message}`,
        });
      }

      return successful;
    }

    return Promise.resolve(false);
  };

  /**
   * Allows the user to select a endpoint from the endpoints array, logout the user from the current
   * selected endpoint
   */
  const selectEndpoint = async (id: string): Promise<Endpoint | undefined> => {
    const endpoint = getEndpointById(id);
    if (endpoint && endpoint.url) {
      if (selectedEndpoint.authToken) {
        await logout();
      }

      selectedEndpoint = endpoint;
      allApiPaths = createApiPaths({ version: endpoint.version });
      apiPathsAuthenticator = allApiPaths.authenticator;

      return endpoint;
    }
    return undefined;
  };

  // Check if the server has noauth enabled
  const checkNoAuth = async (): Promise<null> => {
    let resp;

    try {
      resp = await QorusRequest.get({ path: `${apiPathsAuthenticator.validateNoAuth}` });
      const _noauth = resp.data.noauth;

      if (typeof _noauth === 'boolean') {
        noauth = _noauth;
      }

      return null;
    } catch (error: any) {
      logger.log({
        level: 'error',
        message: `Couldn't check the noauth status: ${error.statusCode}, ErrorMessage: ${error.message}`,
      });
    }

    return null;
  };

  /**
   * A getter to return selected {@link Endpoint}
   */
  const getSelectedEndpoint = (): Endpoint | undefined => {
    return selectedEndpoint;
  };

  /**
   * Validates the local stored authentication token for the endpoint
   */
  const validateLocalUserToken = async (endpointId: string): Promise<string | 'invalid' | null> => {
    const authToken = getKeyValLocal(`auth-token-${endpointId}`);

    if (authToken) {
      try {
        const resp = await QorusRequest.get({
          path: `${apiPathsAuthenticator.validateToken}`,
          data: { token: authToken },
        });

        if (typeof resp === 'string') {
          return authToken;
        }
      } catch (error: any) {
        if (error.code === '409') {
          return 'invalid';
        } else
          logger.log({
            level: 'error',
            message: `Can't validate token, ErrorCode: ${error.code}, ErrorMessage: ${error.message}`,
          });
      }
    }

    return null;
  };

  /**
   * Takes optional username and password parameters to authenticate the user.
   If the username and password is not provided it tries to authenticate using the locally stored token from the selected {@link Endpoint}
   */
  const login = async (loginConfig: LoginParams): Promise<string | undefined> => {
    if (!noauth) {
      const { user, pass } = loginConfig;
      const { id } = selectedEndpoint;
      const currentUserToken = await validateLocalUserToken(id);

      if (!(user && pass) && !currentUserToken) {
        logger.log({ level: 'error', message: 'Authentication is required with Username and Password' });

        return undefined;
      }

      if (currentUserToken && currentUserToken !== 'invalid') {
        selectedEndpoint.authToken = currentUserToken;
        return currentUserToken;
      } else {
        const resp = await QorusRequest.post({
          path: `${apiPathsAuthenticator.login}`,
          data: { user, pass },
        });
        const responseData = resp as AxiosResponse;
        const error = resp as unknown as AxiosError;

        if (error?.code) {
          logger.log({ level: 'error', message: `Couldn't sign in user ${error.code} ${error.message}` });
          return undefined;
        }
        const { token } = responseData?.data;
        selectedEndpoint.authToken = token;
        setKeyValLocal({ key: `auth-token-${id}`, value: token });

        return token;
      }
    }
    return undefined;
  };

  /**
   * Allows the user to add/initialize a new endpoint
   */
  const initEndpoint = async (props: InitEndpoint): Promise<Endpoint> => {
    const { id, url, version, user, pass } = props;
    const newEndpoint: Endpoint = {
      url,
      id,
      version: version ? version : 'latest',
    };
    const endpoint = getEndpointById(id);

    if (!endpoint) {
      endpoints.push(newEndpoint);

      if (selectedEndpoint) {
        await selectEndpoint(id);
      } else {
        selectedEndpoint = newEndpoint;
      }

      await checkNoAuth();
      if (user && pass) await login({ user, pass });
      return selectedEndpoint;
    } else {
      if (selectedEndpoint) {
        await selectEndpoint(id);
      } else {
        selectedEndpoint = newEndpoint;
      }

      await checkNoAuth();
      if (user && pass) await login({ user, pass });
      return selectedEndpoint;
    }
  };

  /**
   * Allows the user to renew the selected endpoint authentication token
   */
  const renewSelectedEndpointToken = async (props: LoginParams): Promise<string | undefined> => {
    const { user, pass } = props;

    let token;
    if (selectedEndpoint) {
      token = await login({ user, pass });
    }
    if (typeof token == 'string') return token;

    return undefined;
  };

  /**
   * A getter to return the auth token of the selected {@link Endpoint}
   */
  const getAuthToken = (): string | undefined => {
    return selectedEndpoint.authToken;
  };

  /**
   * A getter to get the api {@link Version} of a {@link Endpoint}
   */
  const getEndpointVersion = (id?: string): Version | undefined => {
    if (id) {
      return getEndpointById(id)?.version;
    } else {
      if (selectedEndpoint) {
        return selectedEndpoint.version;
      }
    }

    return undefined;
  };

  /**
   * A setter to edit the {@link Version} of the {@link Endpoint}
   */
  const setEndpointVersion = async (
    version: Version,
    id: string = selectedEndpoint.id,
  ): Promise<Version | undefined> => {
    if (id) {
      const endpoint = getEndpointById(id);

      if (endpoint && endpoint.id && endpoint.version) {
        endpoints[endpoints.indexOf(endpoint)].version = version;

        if (selectedEndpoint && selectedEndpoint.id === endpoint.id) {
          selectedEndpoint.version = version;
          allApiPaths = createApiPaths({ version });
          apiPathsAuthenticator = allApiPaths.authenticator;
          await logout();
        }

        logger.log({ level: 'info', message: 'Changed endpoint version successfully.' });
        return version;
      }

      logger.log({ level: 'info', message: 'enpoint does not exist, please try again.' });
      return undefined;
    }

    logger.log({ level: 'info', message: 'id is required to change the version of a endpoint.' });
    return undefined;
  };

  /**
   * A setter to set the url of the selected {@link Endpoint}
   */
  const setEndpointUrl = async (url: string, id: string = selectedEndpoint.id): Promise<string | undefined> => {
    if (id) {
      const endpoint = getEndpointById(id);

      if (endpoint && endpoint.id && endpoint.url) {
        endpoints[endpoints.indexOf(endpoint)].url = url;

        if (selectedEndpoint && selectedEndpoint.id === endpoint.id) {
          selectedEndpoint.url = url;
          await logout();
        }

        logger.log({ level: 'info', message: 'Changed endpoint url successfully.' });
        return url;
      }

      logger.log({ level: 'info', message: 'enpoint does not exist, please try again.' });
      return undefined;
    }

    logger.log({ level: 'info', message: 'id is required to change the url of a endpoint.' });
    return undefined;
  };

  /**
   * A getter to return the api paths for the selected {@link Endpoint}
   */
  const getApiPaths = (): ApiPaths => {
    return allApiPaths;
  };

  /**
   * A getter to get all the available {@link Endpoints}
   */
  const getAllEndpoints = (): Endpoint[] => {
    return endpoints;
  };

  return {
    initEndpoint,
    login,
    logout,
    getAuthToken,
    setEndpointUrl,
    selectEndpoint,
    getSelectedEndpoint,
    renewSelectedEndpointToken,
    getEndpointById,
    getEndpointVersion,
    setEndpointVersion,
    getApiPaths,
    getAllEndpoints,
  };
};

export const QorusAuthenticator: Authenticator = _QorusAuthenticator();
