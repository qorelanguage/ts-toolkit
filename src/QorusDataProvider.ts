import { AxiosResponse } from 'axios';
import ErrorAxios, { ErrorAxiosParams } from './managers/error/ErrorAxios';
import logger from './managers/logger';
import { QorusOptions } from './QorusOptions';
import QorusRequest from './QorusRequest';
import { apiPathsInitial } from './utils/apiPaths';

const getRequestPath = (path: string[]) => {
  let requestPath = '';
  path.forEach((pth) => {
    requestPath = requestPath + pth + '/';
  });
  return requestPath;
};

export type Context = 'record' | 'api' | 'event' | 'message' | 'type' | undefined;

/** Constructor Options for the data provider */
export type ConstructorOptions = any;

/** Get request response data from DataProvider api */
export type ResponseData = any;

/** Get request error data from DataProvider api */
export type ResponseError = any;

/** Qorus DataProvider data object */
export type ProviderData = any;

export class QorusDataProvider {
  /** Array of path strings, linking to the current provider path extension */
  path?: string[] = [];

  /** Get Request response data for a data provider  */
  responseData: ResponseData;

  /** Current context for the data provider */
  context: Context;

  /** Data provider data with children object */
  providerData?: ProviderData;

  /** Get Request error data if error received */
  responseError?: ResponseError;

  constructor(options?: {
    path: string[];
    responseData: ResponseData;
    context: Context;
    providerData?: ProviderData;
    responseError?: ResponseError;
  }) {
    if (options) {
      this.path = options.path;
      this.responseData = options.responseData;
      this.context = options.context;
      this.providerData = options.providerData;
      this.responseError = options.responseError;
    }
  }

  private async fetchWithContext(context: Context) {
    // Making a put request
    const requestPath = getRequestPath([apiPathsInitial.dataProviders.browse]);
    const requestData = { context: context };

    const result = await QorusRequest.put({
      path: requestPath,
      data: requestData,
    });

    const response = result as AxiosResponse;
    const error = result as ErrorAxiosParams;

    if (error.status) {
      throw new ErrorAxios(error);
    }

    const responseData = response?.data;
    const responseError = error.desc;

    return new QorusDataProvider({
      path: [apiPathsInitial.dataProviders.browse],
      responseData: responseData,
      context,
      providerData: responseData,
      responseError,
    });
  }

  /**
   * Get record of Data Providers with context 'record' from /dataprovider/browse endpoint
   * @returns Array of DataProviders records
   */
  async getRecord(): Promise<QorusDataProvider> {
    return this.fetchWithContext('record');
  }

  /**
   * Get record of Data Providers with context 'api'  from /dataprovider/browse endpoint
   * @returns array of DataProviders records
   */

  async getApi(): Promise<QorusDataProvider> {
    return this.fetchWithContext('api');
  }

  /**
   * Get record of Data Providers with context 'Event'  from /dataprovider/browse endpoint
   * @returns Array of DataProviders records
   */
  async getEvent(): Promise<QorusDataProvider> {
    return this.fetchWithContext('event');
  }

  /**
   * Get record of Data Providers with context 'message'  from /dataprovider/browse endpoint
   * @returns Array of DataProviders records
   */
  async getMessage(): Promise<QorusDataProvider> {
    return this.fetchWithContext('message');
  }

  /**
   * Get record of Data Providers with context 'type'  from /dataprovider/browse endpoint
   * @returns Array of DataProviders records
   */
  async getType(): Promise<QorusDataProvider> {
    return this.fetchWithContext('type');
  }

  /**
   * Checks if the children exist on the provider
   * @param name Name of the children you want to find
   * @returns True if the children exist, False otherwise
   */
  has(name: string): boolean {
    const children = this.responseData?.children.find((child) => child.name === name);
    if (typeof children === 'undefined' || !children.name) return false;
    else return true;
  }

  /**
   * A getter to the the stored path array for the current provider
   * @returns Array of path strings
   */
  getPath(): string[] | undefined {
    return this.path;
  }

  /**
   * A getter to get request path for the current provider
   * @param path Optional path array to generate request path
   * @returns Request path string
   */
  getFinalPath(path?: string[]): string {
    return getRequestPath(path ? path : this.path!);
  }

  /**
   * Setter to set path for the current provider
   * @param path Array of path strings to replace for path of the current provider
   *
   */
  setPath(path: string[]) {
    this.path = path;
  }

  /**
   * A getter to get available data for the current provider
   * @returns responseData, providerData and errorData for the current provider
   */
  getData() {
    return { responseData: this.responseData, providerData: this.providerData, errorData: this.responseError };
  }

  /**
   * A getter to get the context for the current provider
   * @returns Context for the api ex: "record";
   */
  getContext(): Context {
    return this.context;
  }

  /**
   * Method to verify if the current provider has children
   * @returns true if the children exist, false otherwise
   */
  hasData(): boolean {
    if (this.responseData.matches_context) return true;
    else return false;
  }

  /**
   * A getter to get available children for the current provider
   * @returns A list of DataProvider children
   */
  getChildren() {
    return this.responseData?.children;
  }

  /**
   * A getter to get children names for the current provider
   * @returns list of children names
   */
  getChildrenNames() {
    /*eslint-disable */
    const children = this.getChildren();
    let names: any = {};
    children?.forEach((child) => (names[child.name] = child.name));

    return names;
  }

  /**
   * A getter to get options by name for a children provider
   * @param childrenName name of the children provider
   * @returns QorusOptions object for the data provider children
   */
  getOptions(childrenName: string): QorusOptions | undefined {
    const children = this.getChildren();
    const filteredChildren = children?.filter((child) => child.name === childrenName);

    if (!filteredChildren || !filteredChildren[0]) {
      logger.error(`Children for the provider "${childrenName}" does not exist, please verify if the provider exist`);
      return;
    }
    return new QorusOptions(filteredChildren[0]);
  }

  /**
   * A getter to get options by name for a children provider
   * @returns QorusOptions object array
   */
  getAllOptions(): QorusOptions[] {
    const children = this.getChildren();
    let allOptions: any[] = [];
    children.forEach((child) => {
      const option = new QorusOptions(child);
      allOptions.push(option);
    });
    return allOptions;
  }

  /**
   * Method to select the next children from the current provider for further operations
   * @param select next children to be selected
   * @param QorusDataProvider constructor options for the next children
   * @returns {@link QorusDataProvider} new object
   */
  async get(select?: string, providerOptions?: ProviderData) {
    if (!select) return this;

    const childData = await fetchProvider(this, this.context, select, providerOptions);
    return childData;
  }
}

/**
 * fetchProvider creates a put request to the QorusServer with context and also adds the
 * functionality to select the next children from dataprovider response list
 * @param obj previous parent object
 * @param context context parameter for the request
 * @param select next provider to be selected, chosen from the parent's children list
 * @param QorusDataProvider constructor options for the dataprovider to be selected next
 * @returns ProviderWithOptions object
 */
const fetchProvider = async (obj: QorusDataProvider, context: Context, select?: string, providerOptions?: any) => {
  const children = obj.getChildren();
  const _path = obj.getPath();

  if (select && _path) {
    if (_path[_path.length - 1] !== select) {
      _path.push(select);
    }
  }

  const requestPath = getRequestPath(_path!);
  const requestData = { context: context, provider_options: providerOptions ?? {} };
  // Making a put request
  const result = await QorusRequest.put({
    path: requestPath,
    data: requestData,
  });

  const response = result as AxiosResponse;
  const error = result as ErrorAxiosParams;

  if (error.status) {
    throw new ErrorAxios(error);
  }
  const providerData = children?.filter((object) => object.name === select);
  const providerResponse = response?.data;
  const responseError = error.desc;

  return new QorusDataProvider({ path: _path!, responseData: providerResponse, context, providerData, responseError });
};

export default new QorusDataProvider();
