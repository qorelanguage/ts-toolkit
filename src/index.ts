export * as Authenticator from './QorusAuthenticator';
export * as HttpRequest from './QorusRequest';
export * as DataProvider from './QorusDataProvider';
export * as QorusOptions from './QorusOptions';
export * as Validator from './QorusValidator';

export { default as QorusAuthenticator } from './QorusAuthenticator';
export { default as QorusRequest } from './QorusRequest';
export { default as QorusDataProvider } from './QorusDataProvider';
export { default as QorusValidator } from './QorusValidator';

export { Version, ApiPaths, ApiPathsAuthenticator, ApiPathsDataProvider } from './utils/apiPaths';
export { Context } from './QorusDataProvider';
export interface ObjectWithStringKey {
  [x: string]: string;
}
export interface ObjectWithAnyKey {
  [x: string]: any;
}
