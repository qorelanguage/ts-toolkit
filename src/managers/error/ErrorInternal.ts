import BaseError from './Error';
import { errorCodes } from './errorCodes';

class ErrorInternal extends BaseError {
  constructor(
    description: string,
    name = errorCodes.INTERNAL.name,
    statusCode = errorCodes.INTERNAL.code,
    isOperational = true,
  ) {
    super(description, isOperational, name, statusCode);
  }
}

export default ErrorInternal;
