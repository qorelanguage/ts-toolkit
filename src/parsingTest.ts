import { AddEndpoint, LoginParams } from './QorusAuthenticator';

/**
 * Checks the validity of the selected endpoint, if the endpoint data ar valid returns true, false otherwise.
 * @param data {@link AddEndpoint} to be checked
 * @param withCredentials boolean to check if the endpoint has credentials
 * @returns boolean true if the endpoint data is valid, false otherwise
 */
const validateEndpointData = (data: AddEndpoint & LoginParams, withCredentials?: boolean): boolean => {
  return true;
};
