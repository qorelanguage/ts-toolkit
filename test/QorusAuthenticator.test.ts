import dotenv from 'dotenv';
import { QorusAuthenticator } from '../src';
import ErrorInternal from '../src/managers/error/ErrorInternal';
import ErrorQorusRequest from '../src/managers/error/ErrorQorusRequest';

dotenv.config();

//const loggerMock = jest.spyOn(logger, 'log');

if (!(process.env.ENDPOINT && process.env.TESTUSER && process.env.TESTPASS)) {
  throw new Error('Missing required environment variables');
}

describe('QorusLogin Utility Class Tests', () => {
  jest.setTimeout(30000);
  it('Should initialize the endpoint and assign it to the selected endpoint', () => {
    QorusAuthenticator.addEndpoint({ url: process.env.ENDPOINT!, endpointId: 'rippy' });

    const endpoint = QorusAuthenticator.getSelectedEndpoint();
    expect(endpoint).toMatchSnapshot();
  });

  it('Should return user token after authentication (login)', async () => {
    let token = await QorusAuthenticator.login({ user: process.env.TESTUSER!, pass: process.env.TESTPASS! });

    expect(typeof token).toEqual('string');
  });

  it('Should return the enpoint from the endpoints array', () => {
    const endpoint = QorusAuthenticator.getEndpointById('rippy');

    expect(endpoint?.endpointId).toEqual('rippy');
  });

  it('Should return all the available endpoints ', () => {
    const endpoints = QorusAuthenticator.getAllEndpoints();

    expect(endpoints).not.toBeNull();
  });

  it('Should return version of the selected endpoint', () => {
    const version = QorusAuthenticator.getEndpointVersion();

    expect(version).toEqual('latest');
  });

  it('Should return api paths for the selected endpoint', () => {
    expect(QorusAuthenticator.getApiPaths()).toMatchSnapshot();
  });

  it('Should return current user token if the user is authenticated', () => {
    const token = QorusAuthenticator.getAuthToken();

    expect(typeof token).toEqual('string');
  });

  it('Should return the current endpoint', () => {
    const config = QorusAuthenticator.getSelectedEndpoint();

    expect(config!.endpointId).toMatchSnapshot();
  });

  it('Should return all the endpoints', () => {
    const endpoints = QorusAuthenticator.getAllEndpoints();

    expect(endpoints.length).toMatchSnapshot();
  });
});

describe('QorusLogin Utility Error Tests', () => {
  it('Should throw an Authentication Error when user tries to authenticate with wrong credentials', async () => {
    let err;
    try {
      if (process.env.ENDPOINT) QorusAuthenticator.addEndpoint({ url: process.env.ENDPOINT, endpointId: 'rippy2' });
      await QorusAuthenticator.login({ user: 'bob', pass: 'pass' });
    } catch (error) {
      err = error;
    }

    expect(err instanceof ErrorQorusRequest).toEqual(true);
  });

  it('Should throw an error if the user does not provide username and password for authentication.', async () => {
    let err;
    try {
      if (process.env.ENDPOINT) QorusAuthenticator.addEndpoint({ url: process.env.ENDPOINT, endpointId: 'rippy2' });
      await QorusAuthenticator.login();
    } catch (error) {
      err = error;
    }
    expect(err instanceof ErrorInternal).toEqual(true);
  });

  it('Should throw an Internal Error if the id and url are not valid to initialize an endpoint', async () => {
    let err;
    try {
      if (process.env.ENDPOINT) QorusAuthenticator.addEndpoint({ url: '', endpointId: '' });
    } catch (error) {
      err = error;
    }
    expect(err instanceof ErrorInternal).toEqual(true);
  });

  it('Should throw an Internal error if the no-auth status cannot be checked', async () => {
    let err;
    try {
      if (process.env.ENDPOINT)
        QorusAuthenticator.addEndpoint({
          url: 'https://sandbox.qoretechnologies.',
          endpointId: 'rippy2',
        });
      await QorusAuthenticator.login({
        user: 'some',
        pass: 'some',
      });
    } catch (error) {
      err = error;
    }
    expect(err instanceof ErrorInternal).toEqual(true);
  });

  it('Should throw an Internal error if url is not valid for initializing endpoint', async () => {
    let err;
    try {
      if (process.env.ENDPOINT)
        QorusAuthenticator.addEndpoint({
          url: '',
          endpointId: 'rippy2',
        });
    } catch (error) {
      err = error;
    }
    expect(err instanceof ErrorInternal).toEqual(true);
  });

  it('Should throw an Internal error if id is not valid for initializing endpoint', async () => {
    let err;
    try {
      if (process.env.ENDPOINT)
        QorusAuthenticator.addEndpoint({
          url: 'https://sandbox.qoretechnologies.com',
          endpointId: '',
        });
    } catch (error) {
      err = error;
    }
    expect(err instanceof ErrorInternal).toEqual(true);
  });

  it('Should throw an Internal error if id and url is not valid for initializing endpoint', async () => {
    let err;
    try {
      if (process.env.ENDPOINT)
        QorusAuthenticator.addEndpoint({
          url: '',
          endpointId: '',
        });
    } catch (error) {
      err = error;
    }
    expect(err instanceof ErrorInternal).toEqual(true);
  });
});
