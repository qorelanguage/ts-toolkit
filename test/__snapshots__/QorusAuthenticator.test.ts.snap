// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QorusLogin Utility Class Tests Should initialize the endpoint and assign it to the selected endpoint 1`] = `
{
  "endpointId": "rippy",
  "url": "https://sandbox.qoretechnologies.com",
  "version": "latest",
}
`;

exports[`QorusLogin Utility Class Tests Should return all the endpoints 1`] = `1`;

exports[`QorusLogin Utility Class Tests Should return api paths for the selected endpoint 1`] = `
{
  "authenticator": {
    "login": "/api/latest/public/login",
    "logout": "/api/latest/logout",
    "validateNoAuth": "/api/latest/public/info",
    "validateToken": "/api/latest/system?action=validateWsToken",
  },
  "dataProviders": {
    "browse": "/api/latest/dataprovider/browse",
  },
}
`;

exports[`QorusLogin Utility Class Tests Should return the current endpoint 1`] = `"rippy"`;
